x/auction/keeper/msg_server_submit_sealed_bid.go

```diff
@@ -0,0 +1,92 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdkmath "cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/spf13/cast"
+
+	"github.com/Stride-Labs/stride/v5/x/auction/types"
+)
+
+// Get the name of the target pool, if the end-block is already before the current blockchain height, then we can start a new auction
+
+func (k msgServer) SubmitSealedBid(goCtx context.Context, msg *types.MsgSubmitSealedBid) (*types.MsgSubmitSealedBidResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	now := cast.ToUint64(ctx.BlockHeight())
+
+	ctx.Logger().Info(fmt.Sprintf("[auction]"))
+	ctx.Logger().Info(fmt.Sprintf("[auction] Sealed bid incoming %s at block %d", msg.GetHashedBid(), now))
+
+	auctionPools := k.Keeper.GetAllAuctionPool(ctx)
+	for _, pool := range auctionPools {
+		if pool.GetId() == msg.GetPoolID() {
+			auction := pool.GetLatestAuction()
+			if auction != nil && auction.GetAlgorithm() == types.AuctionType_SEALEDBID {
+				sealedAuction := auction.GetAuction().(*types.SealedBidAuction)
+				//ctx.Logger().Info(fmt.Sprintf("[auction] Sealed Auction %+v", sealedAuction))
+				if sealedAuction != nil && sealedAuction.GetStatus() == types.AuctionState_RUNNING {
+					creator := msg.GetCreator()
+					sealedValue := msg.GetHashedBid()
+
+					// Take in collateral which will be refunded at the end iff they reveal to avoid sybil attack
+					// If they submit multiple bids from same address to overwrite, we should only take on the first one
+
+					// check if they already have a sealed bid
+					alreadySealedBidForCreator := false
+					sealedAuction.GetSealedBids()
+					for _, b := range sealedAuction.GetSealedBids() {
+						if b.Address == creator {
+							alreadySealedBidForCreator = true
+						}
+					}
+
+					var collateralErr error
+					if alreadySealedBidForCreator {
+						collateralErr = nil
+					} else {
+						// Send the collateral to the pool address if this is first sealed message from bidder address
+						amount := sealedAuction.GetAuctionProperties().Collateral
+						colDenom := pool.GetPoolProperties().BidDenom
+						coin := sdk.NewCoin(colDenom, sdkmath.NewIntFromUint64(amount))
+						coins := sdk.NewCoins(coin)
+						bidderAddress, _ := sdk.AccAddressFromBech32(creator)
+						poolAddress, _ := sdk.AccAddressFromBech32(pool.GetPoolProperties().PoolAddress)
+
+						if k.bankKeeper.BlockedAddr(poolAddress) {
+							ctx.Logger().Info(fmt.Sprintf("[auction] pool address is blocked!"))
+						}
+						if k.bankKeeper.BlockedAddr(bidderAddress) {
+							ctx.Logger().Info(fmt.Sprintf("[auction] bidder address is blocked!"))
+						}
+
+						ctx.Logger().Info(fmt.Sprintf("[auction] About to send collateral %+v %+v %+v", coin, bidderAddress, poolAddress))
+						collateralErr = k.bankKeeper.SendCoins(ctx, bidderAddress, poolAddress, coins)
```

Comment:
I would change this structure to (pseudocode) ```
if collateralErr, print Err and return```
e.g. a general flow of 
"Assume the success case, handle all failure cases immediately, then assume the next success case, etc."